src/codegen.cpp

```diff
@@ -4578,59 +4562,59 @@ static Function *emit_function(jl_lambda_info_t *lam)
-            if (do_coverage)
-                coverageVisitLine(filename, lno);
-            ctx.lineno = lno;
-        }
-        else if (jl_is_expr(stmt) && ((jl_expr_t*)stmt)->head == line_sym) {
-            lno = jl_unbox_long(jl_exprarg(stmt, 0));
-            #ifdef LLVM37
+            jl_sym_t *file = jl_linenode_file(stmt);
+
+#           ifdef LLVM37
             DIFile *dfil = NULL;
-            #else
+#           else
             MDNode *dfil = NULL;
-            #endif
-            if (jl_array_dim0(((jl_expr_t*)stmt)->args) > 1) {
-                jl_value_t *a1 = jl_exprarg(stmt,1);
-                if (jl_is_symbol(a1)) {
-                    jl_sym_t *file = (jl_sym_t*)a1;
-                    // If the string is not empty
-                    if (*file->name != '\0') {
-                        #ifdef LLVM37
-                        std::map<jl_sym_t *, DIFile *>::iterator it = filescopes.find(file);
-                        #else
-                        std::map<jl_sym_t *, MDNode *>::iterator it = filescopes.find(file);
-                        #endif
-                        if (it != filescopes.end()) {
-                            dfil = it->second;
-                        }
-                        else {
-                            #ifdef LLVM37
-                            dfil = filescopes[file] = (DIFile*)dbuilder.createFile(file->name, ".");
-                            #else
-                            dfil = filescopes[file] = (MDNode*)dbuilder.createFile(file->name, ".");
-                            #endif
-                        }
-                    }
+#           endif
+
+            // If the string is not empty
+            if (*file->name != '\0') {
+#               ifdef LLVM37
+                std::map<jl_sym_t *, DIFile *>::iterator it = filescopes.find(file);
+#               else
+                std::map<jl_sym_t *, MDNode *>::iterator it = filescopes.find(file);
+#               endif
+                if (it != filescopes.end()) {
+                    dfil = it->second;
+                } else {
+#                   ifdef LLVM37
+                    dfil = (DIFile*)dbuilder.createFile(file->name, ".");
+#                   else
+                    dfil = (MDNode*)dbuilder.createFile(file->name, ".");
+#                   endif
                 }
             }
+            DebugLoc loc;
+            MDNode *funcscope = (MDNode*)dbuilder.createLexicalBlockFile(SP,topfil);
             if (ctx.debug_enabled) {
                 MDNode *scope;
-                if (dfil == NULL)
-                    scope = SP;
-                else {
+                if ((dfil == topfil || dfil == NULL) &&
+                    lno >= ctx.lineno) // if we are in the top-level file
+                                       // and the current lineno is less than
+                                       // the last, this is same-file inline
```

Comment:
added a TODO note. I also though about indicating entry/exit from inlining with negative line numbers but I don't understand the lowering passes well enough to do that myself yet.